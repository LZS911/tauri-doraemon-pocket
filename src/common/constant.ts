const CONSTANT = {
  TOKEN: '__gl_ui_tvl__',
  THEME_MODE: 't_m',
  COLOR_SCHEME: 'p_m',
  SIDER_WIDTH_WHEN_EXPAND: 200,
  SIDER_WIDTH: 60,
  LANGUAGE_STORAGE_KEY: 'LANGUAGE',
  FONT_FAMILY: 'f_t',
  RESPONSE_SUCCESS_CODE: 0,
  REDIRECT_KEY_PARAMS_NAME: 'target',
};

export default CONSTANT;
