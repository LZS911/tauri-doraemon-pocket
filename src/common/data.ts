export const FormLayout = {
  labelCol: {
    md: { span: 5 },
  },
  wrapperCol: {
    md: { span: 10, offset: 1 },
  },
};
