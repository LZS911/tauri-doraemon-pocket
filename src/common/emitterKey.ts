type EmitterKey = 'storageEvent' | 'newListener' | 'removeListener';

export default EmitterKey;
