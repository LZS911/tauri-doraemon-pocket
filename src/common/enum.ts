export enum ThemeModeEnum {
  'Dark' = 'dark',
  'Light' = 'light',
  'System' = 'system',
}
export enum ColorSchemeEnum {
  'Blue' = 'theme-blue',
  'Purple' = 'theme-purple',
  'Green' = 'theme-green',
}
export enum DarkColorSchemeEnum {
  'Blue' = 'dark-theme-blue',
  'Purple' = 'dark-theme-purple',
  'Green' = 'dark-theme-green',
}

export enum FontFamilyEnum {
  Default = 'Default',
  Inter = 'inter',
  Roboto = 'roboto',
  Poppins = 'poppins',
  Serif = 'serif',
  Mono = 'mono',
  PublicSans = 'publicSans',
}

export enum AutoUpdateEnum {
  Prompt = 'Prompt',
  Silent = 'Silent',
}

export enum ResponseCode {
  SUCCEEDED,
  FAILED,
}
