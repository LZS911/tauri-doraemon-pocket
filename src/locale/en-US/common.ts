export default {
  chinese: 'chinese',
  english: 'english',
  date: {
    month: 'Month',
    week: 'Week',
    monthData: {
      JAN: 'JAN',
      Jan: 'Jan',
      Feb: 'Feb',
      Mar: 'Mar',
      Apr: 'Apr',
      May: 'May',
      Jun: 'Jun',
      Jul: 'Jul',
      Aug: 'Aug',
      Sep: 'Sep',
      Oct: 'Oct',
      Nov: 'Nov',
      Dec: 'Dec',
      Tue: 'Tue',
      Wed: 'Wed',
      Thu: 'Thu',
      Fri: 'Fir',
      Sat: 'Sat',
      Sun: 'Sun',
    },
    weekData: {
      Mon: 'Mon',
    },
  },
  request: {
    noticeFailTitle: 'Request Failed',
  },
  unknownError: 'Unknown Error ',
  title: 'Doraemon Pocket',
};
