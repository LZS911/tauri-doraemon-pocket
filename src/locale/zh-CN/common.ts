export default {
  chinese: '中文',
  english: '英文',
  date: {
    month: '月',
    week: '周',
    monthData: {
      JAN: '一月',
      Jan: '一月',
      Feb: '二月',
      Mar: '三月',
      Apr: '四月',
      May: '五月',
      Jun: '六月',
      Jul: '七月',
      Aug: '八月',
      Sep: '九月',
      Oct: '十月',
      Nov: '十一月',
      Dec: '十二月',
      Tue: '周二',
      Wed: '周三',
      Thu: '周四',
      Fri: '周五',
      Sat: '周六',
      Sun: '周日',
    },
    weekData: {
      Mon: '周一',
    },
  },
  request: {
    noticeFailTitle: '请求错误',
  },
  unknownError: '未知错误',
  title: 'Doraemon Pocket',
  yes: '确认',
  no: '取消',
  apply: '应用',
  submit: '提交',
  back: '返回',
  operator: '操作',
  edit: '编辑',
  delete: '删除',
  addSucceeded: '添加成功',
  editSucceeded: '修改成功',
  deleteSucceeded: '删除成功',
  ok: '确认',
  deleteConfirm: '是否确认删除 {{id}}',
};
