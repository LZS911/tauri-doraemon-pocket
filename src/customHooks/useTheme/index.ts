import { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import CONSTANT from '../../common/constant';
import {
  DarkColorSchemeEnum,
  ThemeModeEnum,
  ColorSchemeEnum,
} from '../../common/enum';
import { IReduxState } from '../../store';
import {
  setCurrentThemeMode as _setCurrentThemeMode,
  setCurrentColorScheme as _setCurrentColorScheme,
} from '../../store/userConfig';
import { IColorSchemeKey } from '../../typing/common.type';
import LocalStorageWrapper from '../../utils/LocalStorageWrapper';
import { baseTypeMatch } from 'rust-like-match';

const addClass = (className: string) => {
  if (document.documentElement.classList.contains(className)) {
    return;
  }

  document.documentElement.classList.add(className);
};

const removeClass = (className: string) => {
  if (!document.documentElement.classList.contains(className)) {
    return;
  }

  document.documentElement.classList.remove(className);
};

const themePrimaryToDark = () => {
  const cls = document.documentElement.classList;
  cls.forEach((v) => {
    if (Object.values(ColorSchemeEnum).includes(v as ColorSchemeEnum)) {
      removeClass(v);
      addClass(`dark-${v}`);
    }
  });
};

const themePrimaryToLight = () => {
  const cls = document.documentElement.classList;
  cls.forEach((v) => {
    if (Object.values(DarkColorSchemeEnum).includes(v as DarkColorSchemeEnum)) {
      removeClass(v);
      addClass(v.replace('dark-', ''));
    }
  });
};

export const getCurrentColorSchemeStrings = (
  isDark: boolean,
  currentColorScheme: ColorSchemeEnum
) => {
  const colorSchemeMap = new Map<ColorSchemeEnum, IColorSchemeKey>([
    [
      ColorSchemeEnum.Blue,
      {
        primary: '#1890ff',
        secondary: '#e6f6ff',
        darkPrimary: '#177ddc',
        darkSecondary: '#213752',
      },
    ],
    [
      ColorSchemeEnum.Purple,
      {
        primary: '#7265e6',
        secondary: '#eeedfc',
        darkPrimary: '#6559c7',
        darkSecondary: '#424062',
      },
    ],
    [
      ColorSchemeEnum.Green,
      {
        primary: '#00a854',
        secondary: '#e3f0e7',
        darkPrimary: '#03934c',
        darkSecondary: '#3a5649',
      },
    ],
  ]);

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const colors = colorSchemeMap.get(currentColorScheme)!;

  if (isDark) {
    return [colors.darkPrimary, colors.darkSecondary];
  }
  return [colors.primary, colors.secondary];
};

export const useInitTheme = () => {
  const dispatch = useDispatch();
  useEffect(() => {
    const _setDark = () => {
      addClass(ThemeModeEnum.Dark);
      themePrimaryToDark();
      LocalStorageWrapper.set(CONSTANT.THEME_MODE, ThemeModeEnum.Dark);
      dispatch(_setCurrentThemeMode(ThemeModeEnum.Dark));
    };

    const isDark =
      LocalStorageWrapper.getOrDefault<ThemeModeEnum>(
        CONSTANT.THEME_MODE,
        ThemeModeEnum.Light
      ) === ThemeModeEnum.Dark;
    const colorSchemeLocal = LocalStorageWrapper.getOrDefault<
      ColorSchemeEnum | DarkColorSchemeEnum
    >(CONSTANT.COLOR_SCHEME, ColorSchemeEnum.Blue);
    addClass(colorSchemeLocal);

    if (isDark) {
      _setDark();
    }
  }, [dispatch]);
};

const useTheme = () => {
  const dispatch = useDispatch();
  const { currentThemeMode, currentColorScheme } = useSelector(
    (state: IReduxState) => {
      return {
        currentThemeMode: state.userConfig.currentThemeMode,
        currentColorScheme: state.userConfig.currentColorScheme,
      };
    }
  );
  const isDark = useMemo(
    () => currentThemeMode === ThemeModeEnum.Dark,
    [currentThemeMode]
  );

  const setCurrentThemeMode = (theme: ThemeModeEnum) => {
    dispatch(_setCurrentThemeMode(theme));
  };

  const setCurrentColorScheme = (theme: ColorSchemeEnum) => {
    dispatch(_setCurrentColorScheme(theme));
  };

  const _setDark = () => {
    addClass(ThemeModeEnum.Dark);
    themePrimaryToDark();
    setCurrentThemeMode(ThemeModeEnum.Dark);
  };

  const _setLight = () => {
    removeClass(ThemeModeEnum.Dark);
    themePrimaryToLight();
    setCurrentThemeMode(ThemeModeEnum.Light);
  };

  const _setColorScheme = (mode: keyof typeof ColorSchemeEnum) => {
    if (isDark) {
      Object.keys(DarkColorSchemeEnum).forEach((key) => {
        if (key === mode) {
          addClass(DarkColorSchemeEnum[key]);
        } else {
          removeClass(
            DarkColorSchemeEnum[key as keyof typeof DarkColorSchemeEnum]
          );
        }
      });
    } else {
      Object.keys(ColorSchemeEnum).forEach((key) => {
        if (key === mode) {
          addClass(ColorSchemeEnum[key]);
        } else {
          removeClass(ColorSchemeEnum[key as keyof typeof ColorSchemeEnum]);
        }
      });
    }

    setCurrentColorScheme(ColorSchemeEnum[mode]);
  };

  const changeThemeMode = (themeMode: ThemeModeEnum) => {
    if (themeMode === currentThemeMode) {
      return;
    }
    if (themeMode === ThemeModeEnum.Dark) {
      _setDark();
    } else {
      _setLight();
    }
  };

  const changeColorScheme = (colorScheme: ColorSchemeEnum) => {
    if (colorScheme === currentColorScheme) {
      return;
    }

    baseTypeMatch(colorScheme, {
      [ColorSchemeEnum.Blue]: () => _setColorScheme('Blue'),
      [ColorSchemeEnum.Green]: () => _setColorScheme('Green'),
      [ColorSchemeEnum.Purple]: () => _setColorScheme('Purple'),
    });
  };

  return {
    currentThemeMode,
    currentColorScheme,
    changeThemeMode,
    changeColorScheme,
    isDark,
  };
};

export default useTheme;
