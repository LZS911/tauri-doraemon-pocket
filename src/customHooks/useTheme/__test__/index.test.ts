import { act, renderHook } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import CONSTANT from '../../../common/constant';
import { ColorSchemeEnum, ThemeModeEnum } from '../../../common/enum';
import useTheme, { getCurrentColorSchemeStrings, useInitTheme } from '../index';

jest.mock('react-redux', () => {
  return {
    ...jest.requireActual('react-redux'),
    useDispatch: jest.fn(),
    useSelector: jest.fn(),
  };
});

describe('test useTheme', () => {
  const useSelectorMock: jest.Mock = useSelector as jest.Mock;
  const useDispatchMock: jest.Mock = useDispatch as jest.Mock;
  const dispatchMock = jest.fn();

  beforeEach(() => {
    useDispatchMock.mockImplementation(() => dispatchMock);
    useSelectorMock.mockImplementation((selector) => {
      return selector({
        userConfig: {
          currentThemeMode: ThemeModeEnum.Light,
          currentColorScheme: ColorSchemeEnum.Blue,
        },
      });
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  test('should match snapshot for getCurrentColorSchemeStrings', () => {
    expect(
      getCurrentColorSchemeStrings(true, ColorSchemeEnum.Blue)
    ).toMatchSnapshot();
    expect(
      getCurrentColorSchemeStrings(true, ColorSchemeEnum.Green)
    ).toMatchSnapshot();
    expect(
      getCurrentColorSchemeStrings(true, ColorSchemeEnum.Purple)
    ).toMatchSnapshot();
    expect(
      getCurrentColorSchemeStrings(false, ColorSchemeEnum.Blue)
    ).toMatchSnapshot();
    expect(
      getCurrentColorSchemeStrings(false, ColorSchemeEnum.Green)
    ).toMatchSnapshot();
    expect(
      getCurrentColorSchemeStrings(false, ColorSchemeEnum.Purple)
    ).toMatchSnapshot();
  });

  test('should set theme mode and color scheme use default value', () => {
    renderHook(() => useInitTheme());
    expect(dispatchMock).toBeCalledTimes(0);
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Blue)
    ).toBeTruthy();
    expect(
      document.documentElement.classList.contains(ThemeModeEnum.Dark)
    ).toBeFalsy();
  });

  test('should be expected change theme when theme mode is light', async () => {
    const { result } = renderHook(() => useTheme());
    expect(result.current.currentThemeMode).toBe(ThemeModeEnum.Light);
    expect(result.current.currentColorScheme).toBe(ColorSchemeEnum.Blue);
    expect(result.current.isDark).toBeFalsy();

    await act(() => {
      result.current.changeThemeMode(ThemeModeEnum.Light);
    });

    expect(dispatchMock).toBeCalledTimes(0);
    expect(
      document.documentElement.classList.contains(ThemeModeEnum.Dark)
    ).toBeFalsy();

    await act(() => {
      result.current.changeColorScheme(ColorSchemeEnum.Blue);
    });
    expect(dispatchMock).toBeCalledTimes(0);
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Blue)
    ).toBeTruthy();

    await act(() => {
      result.current.changeColorScheme(ColorSchemeEnum.Green);
    });
    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setCurrentColorScheme',
      payload: ColorSchemeEnum.Green,
    });
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Green)
    ).toBeTruthy();
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Blue)
    ).toBeFalsy();
    dispatchMock.mockClear();

    await act(() => {
      result.current.changeColorScheme(ColorSchemeEnum.Purple);
    });
    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setCurrentColorScheme',
      payload: ColorSchemeEnum.Purple,
    });
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Purple)
    ).toBeTruthy();
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Green)
    ).toBeFalsy();
    dispatchMock.mockClear();

    await act(() => {
      result.current.changeThemeMode(ThemeModeEnum.Dark);
    });

    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setCurrentThemeMode',
      payload: ThemeModeEnum.Dark,
    });
    expect(
      document.documentElement.classList.contains(ThemeModeEnum.Dark)
    ).toBeTruthy();
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Purple)
    ).toBeFalsy();
    expect(
      document.documentElement.classList.contains(
        `dark-${ColorSchemeEnum.Purple}`
      )
    ).toBeTruthy();
    dispatchMock.mockClear();
  });

  test('should set theme mode when default mode is dark', () => {
    localStorage.setItem(CONSTANT.THEME_MODE, ThemeModeEnum.Dark);
    renderHook(() => useInitTheme());
    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setCurrentThemeMode',
      payload: ThemeModeEnum.Dark,
    });
    expect(
      document.documentElement.classList.contains(ThemeModeEnum.Dark)
    ).toBeTruthy();

    localStorage.setItem(CONSTANT.THEME_MODE, ThemeModeEnum.Light);
    document.documentElement.classList.remove(ThemeModeEnum.Dark);
    document.documentElement.classList.remove(`dark-${ColorSchemeEnum.Blue}`);
  });

  test('should be expected change theme when theme mode is dark', async () => {
    useSelectorMock.mockImplementation((selector) => {
      return selector({
        userConfig: {
          currentThemeMode: ThemeModeEnum.Dark,
          currentColorScheme: ColorSchemeEnum.Green,
        },
      });
    });

    const { result } = renderHook(() => useTheme());
    expect(result.current.currentThemeMode).toBe(ThemeModeEnum.Dark);
    expect(result.current.currentColorScheme).toBe(ColorSchemeEnum.Green);
    expect(result.current.isDark).toBeTruthy();

    await act(() => {
      result.current.changeColorScheme(ColorSchemeEnum.Blue);
    });
    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setCurrentColorScheme',
      payload: ColorSchemeEnum.Blue,
    });
    expect(
      document.documentElement.classList.contains(
        `dark-${ColorSchemeEnum.Blue}`
      )
    ).toBeTruthy();
    expect(
      document.documentElement.classList.contains(ColorSchemeEnum.Blue)
    ).toBeFalsy();

    dispatchMock.mockClear();

    await act(() => {
      result.current.changeThemeMode(ThemeModeEnum.Light);
    });

    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setCurrentThemeMode',
      payload: ThemeModeEnum.Light,
    });
    expect(
      document.documentElement.classList.contains(ThemeModeEnum.Dark)
    ).toBeFalsy();
  });
});
