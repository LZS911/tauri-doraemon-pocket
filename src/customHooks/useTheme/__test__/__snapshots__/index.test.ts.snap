// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test useTheme should match snapshot for getCurrentColorSchemeStrings 1`] = `
Array [
  "#177ddc",
  "#213752",
]
`;

exports[`test useTheme should match snapshot for getCurrentColorSchemeStrings 2`] = `
Array [
  "#03934c",
  "#3a5649",
]
`;

exports[`test useTheme should match snapshot for getCurrentColorSchemeStrings 3`] = `
Array [
  "#6559c7",
  "#424062",
]
`;

exports[`test useTheme should match snapshot for getCurrentColorSchemeStrings 4`] = `
Array [
  "#1890ff",
  "#e6f6ff",
]
`;

exports[`test useTheme should match snapshot for getCurrentColorSchemeStrings 5`] = `
Array [
  "#00a854",
  "#e3f0e7",
]
`;

exports[`test useTheme should match snapshot for getCurrentColorSchemeStrings 6`] = `
Array [
  "#7265e6",
  "#eeedfc",
]
`;
